File: app/src/main/assets/modules/__tasks__.js
// noinspection JSUnusedGlobalSymbols,JSUnusedLocalSymbols

/* Overwritten protection. */

let { ui, threads, files } = global;

/**
 * @param {ScriptRuntime} scriptRuntime
 * @param {org.mozilla.javascript.Scriptable | global} scope
 * @return {Internal.Tasks}
 */
module.exports = function (scriptRuntime, scope) {
    const TimedTask = org.autojs.autojs.timing.TimedTask;
    const IntentTask = org.autojs.autojs.timing.IntentTask;
    const TimedTaskManager = org.autojs.autojs.timing.TimedTaskManager;
    const ExecutionConfig = org.autojs.autojs.execution.ExecutionConfig;
    const DynamicBroadcastReceivers = org.autojs.autojs.external.receiver.DynamicBroadcastReceivers;

    let _ = {
        Tasks: (/* @IIFE */ () => {
            /**
             * @implements Internal.Tasks
             */
            let Task = function () {
                /* Empty body. */
            };

            Task.prototype = {
                constructor: Task,
                /**
                 * @template {TimedTask$|IntentTask$|null} T
                 * @param {T} task
                 * @return {T}
                 */
                addTask(task) {
                    if (task) {
                        TimedTaskManager.addTask(task);
                    }
                    return task || null;
                },
                /**
                 * @param {Timers.TimedTask.Daily} [options]
                 * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                addDailyTask(options) {
                    let opt = options || {};

                    let localTime = _.parseDateTime('LocalTime', opt.time || Date.now());
                    let path = _.parsePath(opt.path);
                    let config = _.parseConfig(opt);

                    let timedTask = this.addTask(TimedTask.dailyTask(localTime, path, config));

                    return _.taskFulfilled(timedTask, opt);
                },
                /**
                 * @param {Timers.TimedTask.Weekly} [options]
                 * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                addWeeklyTask(options) {
                    let timeFlag = 0;
                    let opt = options || {};

                    let daysOfWeek = opt.daysOfWeek || [ _.getCurrentDayOfWeek() ];
                    let daysOfWeekList = _.getDaysOfWeekFlatList();

                    for (let i = 0; i < daysOfWeek.length; i += 1) {
                        let dayString = daysOfWeek[i].toString();
                        let dayIndex = daysOfWeekList.indexOf(dayString.toLowerCase()) % 7;
                        if (dayIndex < 0) {
                            throw Error(`Unknown day: ${dayString}`);
                        }
                        timeFlag |= TimedTask.getDayOfWeekTimeFlag(dayIndex + 1);
                    }

                    let localTime = _.parseDateTime('LocalTime', opt.time || Date.now());
                    let flagsNum = Number(new java.lang.Long(timeFlag));
                    let path = _.parsePath(opt.path);
                    let config = _.parseConfig(opt);

                    let timedTask = this.addTask(TimedTask.weeklyTask(localTime, flagsNum, path, config));

                    return _.taskFulfilled(timedTask, opt);
                },
                /**
                 * @param {Timers.TimedTask.Disposable} options
                 * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                addDisposableTask(options) {
                    let opt = options || {};

                    let localDateTime = _.parseDateTime('LocalDateTime', opt.date || Date.now());
                    let path = _.parsePath(opt.path);
                    let config = _.parseConfig(opt);

                    let timedTask = this.addTask(TimedTask.disposableTask(localDateTime, path, config));

                    return _.taskFulfilled(timedTask, opt);
                },
                /**
                 * @param {Timers.IntentTask.Basic} [options]
                 * @return {IntentTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                addIntentTask(options) {
                    let opt = options || {};

                    let intentTask = (function $iiFe() {
                        let intentTask = new IntentTask();
                        intentTask.setScriptPath(_.parsePath(opt.path));
                        if (typeof opt.action === 'string') {
                            intentTask.setAction(opt.action);
                            if (opt.action === DynamicBroadcastReceivers.ACTION_STARTUP) {
                                // @Indecision opt.local
                                intentTask.setLocal(true);
                            }
                        }
                        if (typeof opt.dataType === 'string') {
                            intentTask.setDataType(opt.dataType);
                        }
                        if (typeof opt.local === 'boolean') {
                            // @FinalDecision opt.action
                            intentTask.setLocal(opt.local);
                        }
                        return intentTask;
                    })();

                    let timedTask = this.addTask(intentTask);

                    return _.taskFulfilled(timedTask, opt);
                },
                /**
                 * @param {number} id
                 * @return {TimedTask$}
                 */
                getTimedTask(id) {
                    return TimedTaskManager.getTimedTask(id);
                },
                /**
                 * @param {number} id
                 * @return {IntentTask$}
                 */
                getIntentTask(id) {
                    return TimedTaskManager.getIntentTask(id);
                },
                /**
                 * @param {TimedTask$|IntentTask$|null} task
                 * @return {TimedTask$|IntentTask$}
                 */
                removeTask(task) {
                    if (task) {
                        TimedTaskManager.removeTask(task);
                    }
                    return task;
                },
                /**
                 * @param {number} id
                 * @param {Timers.TimedTask.Extension} [options]
                 * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                removeTimedTask(id, options) {
                    let opt = options || {};
                    let task = this.removeTask(this.getTimedTask(id));
                    return _.taskFulfilled(task, Object.assign(opt, {
                        condition: () => !this.getTimedTask(id),
                    }));
                },
                /**
                 * @param {number} id
                 * @param {Timers.TimedTask.Extension} [options]
                 * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
                 */
                removeIntentTask(id, options) {
                    let opt = options || {};
                    let task = this.removeTask(this.getIntentTask(id));
                    return _.taskFulfilled(task, Object.assign(opt, {
                        condition: () => !this.getIntentTask(id),
                    }));
                },
                /**
                 * @param {TimedTask$|null} task
                 * @return {TimedTask$|null}
                 */
                updateTask(task) {
                    if (!task) {
                        return null;
                    }
                    task.setScheduled(false);
                    TimedTaskManager.updateTask(task);
                    return task;
                },
                /**
                 * @param {{path?:string}} [options]
                 * @return {TimedTask$[]}
                 */
                queryTimedTasks(options) {
                    let opt = options || {};
                    let path = opt.path;
                    let list = TimedTaskManager.getAllTasksAsList().toArray();
                    return path ? list.filter(task => task.getScriptPath() === path) : list;
                },
                /**
                 * @param {{path?:string,action?:string}} [options]
                 * @return {IntentTask$[]}
                 */
                queryIntentTasks(options) {
                    let opt = options || {};
                    let { path, action } = opt;

                    let list = TimedTaskManager.getAllIntentTasksAsList().toArray();

                    if (!path && !action) {
                        return list;
                    }

                    let pathTrigger = task => !path || task.getScriptPath() === path;
                    let actionTrigger = task => !action || task.getAction() === action;

                    return list.filter(task => pathTrigger(task) && actionTrigger(task));
                },
                /**
                 * @param flag {number} -- number from 0 to 127
                 * @return {number[]}
                 */
                timeFlagToDays(flag) {
                    let days = [];
                    let binaryString = Number(flag).toString(2);
                    let currentDayNumber = binaryString.length - 1;
                    for (let i of binaryString) {
                        if (i !== '0') {
                            days.unshift(currentDayNumber);
                        }
                        currentDayNumber -= 1;
                    }
                    return days;
                },
                /**
                 * @param days {number[]}
                 * @return {number}
                 */
                daysToTimeFlag(days) {
                    return Array(7).fill(0).reduce((a, b, i) => {
                        return a + (days.includes(i) ? Math.pow(2, i) : 0);
                    }, 0);
                },
            };

            return Task;
        })(),
        parsePath(path) {
            if (!path) {
                throw 'A path is necessary for a task';
            }
            if (!files.exists(path)) {
                scriptRuntime.console.warn(`Specified path "${path}" doesn't exist`);
            }
            return files.path(path);
        },
        /**
         * @param {any} num
         * @param {number|function():number} [def=0]
         * @returns {number}
         */
        parseNumber(num, def) {
            return typeof num === 'number' ? num : typeof def === 'function' ? def() : def || 0;
        },
        /**
         * @param {Object} config
         * @param {number} [config.delay=0]
         * @param {number} [config.interval=0]
         * @param {number} [config.loopTimes=1]
         * @return {org.autojs.autojs.execution.ExecutionConfig}
         */
        parseConfig(config) {
            let execConfig = new ExecutionConfig();
            execConfig.setDelay(_.parseNumber(config.delay, 0));
            execConfig.setInterval(_.parseNumber(config.interval, 0));
            execConfig.setLoopTimes(_.parseNumber(config.loopTimes, 1));
            return execConfig;
        },
        /**
         * @param {'LocalTime'|'LocalDateTime'} clazz
         * @param {string|Date|number} dateTime
         * @return {org.joda.time.LocalTime|org.joda.time.LocalDateTime}
         */
        parseDateTime(clazz, dateTime) {
            let clz = (/* @IIFE */ () => {
                switch (clazz) {
                    case 'LocalTime':
                        return org.joda.time.LocalTime;
                    case 'LocalDateTime':
                        return org.joda.time.LocalDateTime;
                    default:
                        throw Error('Unknown clazz for parseDateTime');
                }
            })();
            if (typeof dateTime === 'string') {
                return clz.parse(dateTime);
            }
            if (dateTime instanceof Date) {
                return new clz(dateTime.getTime());
            }
            if (typeof dateTime === 'number') {
                return new clz(dateTime);
            }
            throw Error('Unknown dateTime for parseDateTime');
        },
        /**
         * @param {TimedTask$|IntentTask$} task
         * @param {Timers.TimedTask.Extension|{}} [options]
         * @return {TimedTask$|org.autojs.autojs.core.looper.TimerThread|null}
         */
        taskFulfilled(task, options) {
            let opt = options || {};

            let run = () => {
                if (task) {
                    let timeout = 2e3;
                    let interval = 50;
                    let condition = opt.condition || (() => task['id'] > 0);
                    while (timeout > 0) {
                        if (condition()) {
                            break;
                        }
                        sleep(interval);
                        timeout -= interval;
                    }
                }
                return opt.callback ? opt.callback(task) : task || null;
            };

            return opt.isAsync || ui.isUiThread() ? threads.start(run) : run();
        },
        getCurrentDayOfWeek() {
            return new Date().getDay();
        },
        getDaysOfWeekFlatList() {
            return [
                'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday',
                'mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun',
                '一', '二', '三', '四', '五', '六', '日',
                1, 2, 3, 4, 5, 6, 0,
                1, 2, 3, 4, 5, 6, 7,
            ].map(x => typeof x === 'string' ? x : x.toString());
        },
    };

    // noinspection UnnecessaryLocalVariableJS
    /** @type {Internal.Tasks} */
    const tasks = new _.Tasks();

    return tasks;
};

File: app/src/main/java/org/autojs/autojs/runtime/api/augment/tasks/Tasks.kt
package org.autojs.autojs.runtime.api.augment.tasks

import org.autojs.autojs.annotation.RhinoFunctionBody
import org.autojs.autojs.annotation.RhinoRuntimeFunctionInterface
import org.autojs.autojs.execution.ExecutionConfig
import org.autojs.autojs.extension.AnyExtensions.isJsNullish
import org.autojs.autojs.extension.AnyExtensions.jsBrief
import org.autojs.autojs.extension.AnyExtensions.toRuntimePath
import org.autojs.autojs.extension.ArrayExtensions.toNativeArray
import org.autojs.autojs.extension.FlexibleArray
import org.autojs.autojs.extension.ScriptableExtensions.prop
import org.autojs.autojs.extension.ScriptableObjectExtensions.acquire
import org.autojs.autojs.extension.ScriptableObjectExtensions.inquire
import org.autojs.autojs.external.receiver.DynamicBroadcastReceivers
import org.autojs.autojs.runtime.ScriptRuntime
import org.autojs.autojs.runtime.api.augment.Augmentable
import org.autojs.autojs.runtime.api.augment.threads.Threads
import org.autojs.autojs.runtime.exception.WrappedIllegalArgumentException
import org.autojs.autojs.timing.IntentTask
import org.autojs.autojs.timing.TimedTask
import org.autojs.autojs.timing.TimedTaskManager
import org.autojs.autojs.util.RhinoUtils.callFunction
import org.autojs.autojs.util.RhinoUtils.coerceArray
import org.autojs.autojs.util.RhinoUtils.coerceBoolean
import org.autojs.autojs.util.RhinoUtils.coerceIntNumber
import org.autojs.autojs.util.RhinoUtils.coerceLongNumber
import org.autojs.autojs.util.RhinoUtils.coerceObject
import org.autojs.autojs.util.RhinoUtils.coerceString
import org.autojs.autojs.util.RhinoUtils.isUiThread
import org.autojs.autojs.util.RhinoUtils.newNativeArray
import org.autojs.autojs.util.RhinoUtils.newNativeObject
import org.autojs.autojs.util.RhinoUtils.now
import org.joda.time.LocalDateTime
import org.joda.time.LocalTime
import org.mozilla.javascript.BaseFunction
import org.mozilla.javascript.NativeArray
import org.mozilla.javascript.NativeDate
import org.mozilla.javascript.NativeObject
import org.mozilla.javascript.ScriptableObject
import java.time.LocalDate
import kotlin.math.pow

@Suppress("unused", "UNUSED_PARAMETER")
class Tasks(scriptRuntime: ScriptRuntime) : Augmentable(scriptRuntime) {

    override val selfAssignmentFunctions = listOf(
        ::addTask.name,
        ::addDailyTask.name,
        ::addWeeklyTask.name,
        ::addDisposableTask.name,
        ::addIntentTask.name,
        ::getTimedTask.name,
        ::getIntentTask.name,
        ::removeTask.name,
        ::removeTimedTask.name,
        ::removeIntentTask.name,
        ::updateTask.name,
        ::queryTimedTasks.name,
        ::queryIntentTasks.name,
        ::timeFlagToDays.name,
        ::daysToTimeFlag.name,
    )

    companion object : FlexibleArray() {

        private val daysOfWeekFlatList: List<String> = listOf(
            "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday",
            "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun",
            "一", "二", "三", "四", "五", "六", "日",
            1, 2, 3, 4, 5, 6, 0,
            1, 2, 3, 4, 5, 6, 7,
        ).map { "$it".lowercase() }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun addTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Any = ensureArgumentsOnlyOne(args) {
            addTaskRhino(it)
        }

        @JvmStatic
        @RhinoFunctionBody
        fun addTaskRhino(task: Any?): Any {
            when (task) {
                is TimedTask -> TimedTaskManager.addTaskSync(task)
                is IntentTask -> TimedTaskManager.addTaskSync(task)
                else -> listOf(
                    "Argument task ${task.jsBrief()} for tasks.addTask",
                    "must be either TimedTask or IntentTask",
                ).joinToString(" ").let { throw WrappedIllegalArgumentException(it) }
            }
            return task
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun addDailyTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Any? = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            TimedTask.dailyTask(
                parseLocalTime(opt),
                opt.acquire("path") { it.toRuntimePath(scriptRuntime) },
                parseConfig(opt),
            ).let { task -> taskFulfilled(scriptRuntime, addTaskRhino(task), opt) }
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun addWeeklyTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Any? = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            TimedTask.weeklyTask(
                parseLocalTime(opt),
                parseDayOfWeekTimeFlag(opt).toLong(),
                opt.acquire("path") { it.toRuntimePath(scriptRuntime) },
                parseConfig(opt),
            ).let { task -> taskFulfilled(scriptRuntime, addTaskRhino(task), opt) }
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun addDisposableTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Any? = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            TimedTask.disposableTask(
                parseLocalDateTime(opt),
                opt.acquire("path") { it.toRuntimePath(scriptRuntime) },
                parseConfig(opt),
            ).let { task -> taskFulfilled(scriptRuntime, addTaskRhino(task), opt) }
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun addIntentTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Any? = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            IntentTask().apply {
                opt.acquire("path") { it.toRuntimePath(scriptRuntime) }.let {
                    scriptPath = it
                }
                opt.inquire("action") { coerceString(it) }?.let {
                    action = it
                    if (action == DynamicBroadcastReceivers.ACTION_STARTUP) {
                        // @Indecision opt.local
                        isLocal = true
                    }
                }
                opt.inquire("dataType") { coerceString(it) }?.let {
                    dataType = it
                }
                opt.inquire(listOf("isLocal", "local")) { coerceBoolean(it) }?.let {
                    // @FinalDecision opt.action
                    isLocal = it
                }
            }.let { task -> taskFulfilled(scriptRuntime, addTaskRhino(task), opt) }
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun getTimedTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): TimedTask? = ensureArgumentsOnlyOne(args) {
            getTimedTaskRhino(it)
        }

        @JvmStatic
        @RhinoFunctionBody
        fun getTimedTaskRhino(it: Any?): TimedTask? = TimedTaskManager.getTimedTask(coerceLongNumber(it))

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun getIntentTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): IntentTask? = ensureArgumentsOnlyOne(args) {
            getIntentTaskRhino(it)
        }

        @JvmStatic
        @RhinoFunctionBody
        fun getIntentTaskRhino(it: Any?) = TimedTaskManager.getIntentTask(coerceLongNumber(it))

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun removeTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Boolean = ensureArgumentsOnlyOne(args) {
            removeTaskRhino(it)
        }

        @JvmStatic
        @RhinoFunctionBody
        fun removeTaskRhino(it: Any?): Boolean = when {
            it.isJsNullish() -> false
            it is TimedTask -> TimedTaskManager.removeTaskSync(it)
            it is IntentTask -> TimedTaskManager.removeTaskSync(it)
            else -> throw WrappedIllegalArgumentException("Argument task ${it.jsBrief()} is unacceptable for tasks.removeTask")
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun removeTimedTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Boolean = ensureArgumentsOnlyOne(args) {
            val id = coerceIntNumber(it)
            removeTaskRhino(getTimedTaskRhino(id))
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun removeIntentTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Boolean = ensureArgumentsOnlyOne(args) {
            val id = coerceIntNumber(it)
            removeTaskRhino(getIntentTaskRhino(id))
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun updateTask(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Boolean = ensureArgumentsOnlyOne(args) {
            updateTaskRhino(it)
        }

        @JvmStatic
        @RhinoFunctionBody
        fun updateTaskRhino(task: Any?): Boolean = when {
            task.isJsNullish() -> false
            task is TimedTask -> TimedTaskManager.updateTaskSync(task)
            task is IntentTask -> TimedTaskManager.updateTaskSync(task)
            else -> throw WrappedIllegalArgumentException("Argument task ${task.jsBrief()} is unacceptable for tasks.updateTask")
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun queryTimedTasks(scriptRuntime: ScriptRuntime, args: Array<out Any?>): NativeArray = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            val path = opt.inquire("path") { it.toRuntimePath(scriptRuntime) }
            val taskList = TimedTaskManager.allTasksAsList
            val result = path?.let { taskList.filter { it.scriptPath == path } } ?: taskList

            result.toNativeArray()
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun queryIntentTasks(scriptRuntime: ScriptRuntime, args: Array<out Any?>): NativeArray = ensureArgumentsAtMost(args, 1) { argList ->
            val (options) = argList
            val opt = coerceObject(options, newNativeObject())

            val tasks = TimedTaskManager.allIntentTasksAsList

            val path = opt.inquire("path") { it.toRuntimePath(scriptRuntime) }
            val action = opt.inquire("action") { coerceString(it) }

            tasks.filter {
                when {
                    path != null && it.scriptPath != path -> false
                    action != null && it.action != action -> false
                    else -> true
                }
            }.toNativeArray()
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun timeFlagToDays(scriptRuntime: ScriptRuntime, args: Array<out Any?>): NativeArray = ensureArgumentsOnlyOne(args) {
            val flag = coerceIntNumber(it)
            val days = mutableListOf<Int>()
            val binaryString = flag.toString(2)
            var currentDayNumber = binaryString.length - 1
            for (i in binaryString) {
                if (i != '0') {
                    days.add(0, currentDayNumber)
                }
                currentDayNumber--
            }
            days.toNativeArray()
        }

        @JvmStatic
        @RhinoRuntimeFunctionInterface
        fun daysToTimeFlag(scriptRuntime: ScriptRuntime, args: Array<out Any?>): Int = ensureArgumentsOnlyOne(args) {
            val days = coerceArray(it, newNativeArray()).map { ele -> coerceIntNumber(ele) }
            Array(7) { 0 }.foldIndexed(0) { index, acc, _ ->
                acc + if (days.contains(index)) 2.0.pow(index.toDouble()).toInt() else 0
            }
        }

        private fun parseConfig(config: NativeObject) = ExecutionConfig().apply {
            delay = config.inquire("delay", ::coerceLongNumber, 0L)
            interval = config.inquire("interval", ::coerceLongNumber, 0L)
            loopTimes = config.inquire("loopTimes", ::coerceIntNumber, 1)
        }

        private fun parseLocalTime(options: ScriptableObject): LocalTime {
            val dateTime = options.prop("time").takeUnless { it.isJsNullish() } ?: options.prop("date")
            return when {
                dateTime.isJsNullish() -> LocalTime(now())
                dateTime is Number -> LocalTime(dateTime.toLong())
                dateTime is NativeDate -> LocalTime(dateTime.date)
                dateTime is String -> LocalTime.parse(dateTime)
                else -> throw WrappedIllegalArgumentException("Unknown dataTime ${dateTime.jsBrief()}")
            }
        }

        private fun parseLocalDateTime(options: ScriptableObject): LocalDateTime {
            val localDateTime = options.prop("time").takeUnless { it.isJsNullish() } ?: options.prop("date")
            return when {
                localDateTime.isJsNullish() -> LocalDateTime(now())
                localDateTime is Number -> LocalDateTime(localDateTime.toLong())
                localDateTime is NativeDate -> LocalDateTime(localDateTime.date)
                localDateTime is String -> LocalDateTime.parse(localDateTime)
                else -> throw WrappedIllegalArgumentException("Unknown localDateTime ${localDateTime.jsBrief()}")
            }
        }

        private fun parseDayOfWeekTimeFlag(opt: NativeObject): Int {
            var timeFlag = 0

            val daysOfWeek: List<Any?> = opt.inquire("daysOfWeek", ::coerceArray, listOf(LocalDate.now().dayOfWeek.value % 7).toNativeArray())
            daysOfWeek.forEach { element ->
                val eleString = coerceString(element)
                val dayIndex = daysOfWeekFlatList.indexOf(eleString.lowercase()) % 7
                require(dayIndex >= 0) { "Unknown day: $eleString" }
                timeFlag = timeFlag or TimedTask.getDayOfWeekTimeFlag(dayIndex + 1).toInt()
            }
            return timeFlag
        }

        private fun taskFulfilled(scriptRuntime: ScriptRuntime, task: Any, options: NativeObject? = null): Any? {
            val opt = coerceObject(options, newNativeObject())

            var runnableResult: Any? = null

            val r = Runnable {
                val callback = opt.prop("callback")
                runnableResult = when {
                    callback.isJsNullish() -> task
                    else -> {
                        require(callback is BaseFunction) {
                            "Property \"callback\" for options must be a JavaScript Function instead of ${callback.jsBrief()}"
                        }
                        callFunction(callback, opt, opt, arrayOf(task))
                    }
                }
            }

            return when {
                opt.inquire(listOf("isAsync", "async"), ::coerceBoolean, false) || isUiThread() -> {
                    Threads.start(scriptRuntime, arrayOf(r))
                }
                else -> {
                    r.run()
                    runnableResult
                }
            }
        }

    }

}
